File: sample-java-app/src/main/java/com/unissey/samplejavaapp/ui/views/OvalOverlayView.java
package com.unissey.samplejavaapp.ui.views;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.RectF;
import android.graphics.Region;
import android.os.Build;
import android.util.AttributeSet;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.material.color.MaterialColors;

public class OvalOverlayView extends View {

    private RectF faceArea = new RectF();

    private final int progressColor;

    private float recordingProgress = 0;

    private final Paint overlayPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Paint borderPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Paint progressPaint = new Paint(Paint.ANTI_ALIAS_FLAG);

    private final Path facePath = new Path();

    public OvalOverlayView(Context context) {
        this(context, null);
    }

    public OvalOverlayView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);

        progressColor = MaterialColors.getColor(
                this,
                android.R.attr.colorPrimary
        );

        overlayPaint.setStyle(Paint.Style.FILL);

        borderPaint.setStyle(Paint.Style.STROKE);
        borderPaint.setColor(Color.WHITE);
        borderPaint.setStrokeWidth(12);

        progressPaint.setStyle(Paint.Style.STROKE);
        progressPaint.setStrokeWidth(12);
        progressPaint.setStrokeCap(Paint.Cap.ROUND);
    }

    public void setFaceArea(RectF value) {
        faceArea = value;
        invalidate();
    }

    public void setRecordingProgress(float recordingProgress) {
        this.recordingProgress = recordingProgress;
        invalidate();
    }

    /* -------------------- Drawing -------------------- */

    @Override
    protected void onDraw(@NonNull Canvas canvas) {
        super.onDraw(canvas);

        // Create oval path
        facePath.reset();
        facePath.addOval(faceArea, Path.Direction.CW);

        int save = canvas.save();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            canvas.clipOutPath(facePath);
        } else {
            canvas.clipPath(facePath, Region.Op.DIFFERENCE);
        }

        int overlayColor = 0x88000000;
        overlayPaint.setColor(overlayColor);
        canvas.drawRect(
                0,
                0,
                getWidth(),
                getHeight(),
                overlayPaint
        );
        canvas.restoreToCount(save);

        // Draw oval border
        canvas.drawOval(faceArea, borderPaint);

        // Draw progress arc
        progressPaint.setColor(progressColor);
        canvas.drawArc(
                faceArea,
                -90,
                recordingProgress * 360,
                false,
                progressPaint
        );
    }
}

File: sample-legacy-app/src/main/java/com/unissey/samplelegacyapp/ui/views/OvalOverlayView.kt
package com.unissey.samplelegacyapp.ui.views

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import android.graphics.RectF
import android.graphics.Region
import android.os.Build
import android.util.AttributeSet
import android.view.View
import androidx.core.graphics.withSave
import com.google.android.material.color.MaterialColors

class OvalOverlayView(
    context: Context,
    attrs: AttributeSet? = null
) : View(context, attrs) {
    var faceArea: RectF = RectF()
        set(value) {
            field = value
            invalidate()
        }

    var overlayColor: Int = 0x88000000.toInt()
    var progressColor: Int = MaterialColors.getColor(
        this,
        android.R.attr.colorPrimary
    )
    var recordingProgress: Float = 0f

    private val overlayPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        style = Paint.Style.FILL
    }

    private val borderPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        style = Paint.Style.STROKE
        color = Color.WHITE
        strokeWidth = 12f
    }

    private val progressPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        style = Paint.Style.STROKE
        strokeWidth = 12f
        strokeCap = Paint.Cap.ROUND
    }

    private val facePath = Path()

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        // Draw a dimmed overlay with an oval hole
        facePath.reset()
        facePath.addOval(faceArea, Path.Direction.CW)

        canvas.withSave {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                clipOutPath(facePath)
            } else {
                @Suppress("DEPRECATION")
                clipPath(facePath, Region.Op.DIFFERENCE)
            }

            overlayPaint.color = overlayColor
            drawRect(0f, 0f, width.toFloat(), height.toFloat(), overlayPaint)
        }

        // Draw the oval border
        canvas.drawOval(faceArea, borderPaint)

        // Draw the progress arc
        progressPaint.color = progressColor
        canvas.drawArc(
            faceArea,
            -90f,
            recordingProgress * 360f,
            false,
            progressPaint
        )
    }
}